contracts/StorageManager.sol

```diff
@@ -14,87 +15,83 @@ contract StorageManager {
 
     /*
     Offer represents:
-     - availableCapacity: the amount of bytes currently offered. When the capacity is zero, already started Agreement can't be prolonged or re-started
-     - maximumDuration: the maximum time (in seconds) for which a customer can prepay.
+     - availableCapacity: the amount of bytes currently offered. When the capacity is zero, already started Agreement can't be prolonged or re-started ==> it is terminated
      - billingPlans: maps a billing period to a billing price. When a price is 0, the period is not offered.
      - agreementRegistry: the proposed and accepted Agreement
     */
     struct Offer {
         uint128 availableCapacity;
-        uint128 maximumDuration;
         mapping(uint64 => uint64) billingPlans;
         mapping(bytes32 => Agreement) agreementRegistry; // link to agreement that are accepted under this offer
     }
 
     /*
     Agreement represents:
-     - billingPrice: price per byte that is collected per each period. The contract can be cancelled by the proposer every duration seconds since the start.
-     - billingPeriod: period how often are funds collected.
+     - billingPrice: price per byte that is collected per each period.
+     - billingPeriod: period how often billing happens.
      - size: allocated size for the Agreement (in bytes, rounded up)
-     - startDate: when the Agreement was accepted
-     - numberOfPeriodsDeposited: number of periods (chosenPrice.duration seconds) that is deposited in the contracts.
-       At startDate * numberOfPeriodsDeposited seconds the Agreement expires unless topped up in the meantime
-     - numberOfPeriodsWithdrawn how many periods are withdrawn from the numberOfPeriodsDeposited. Provider can withdraw every period seconds since the start
-     - startDate: when Agreement was accepted
+     - availableFunds: funds available for the billing of the Agreement.
+     - lastPayoutDate: When was the last time Provider was payed out. Zero either means non-existing or terminated Agreement.
     */
     struct Agreement {
         uint64 billingPrice;
         uint64 billingPeriod;
-        uint64 numberOfPeriodsDeposited;
-        uint64 numberOfPeriodsWithdrawn;
+        uint256 availableFunds;
         uint128 size;
-        uint128 startDate;
+        uint128 lastPayoutDate;
```

Comment:
@AuHau  can you briefly explain the rationale for "lastPayoutDate"? At the beginning when a new Agreement is created you set it up with that date, but in that case it would not be the lastPayoutDate but the agreement start date? Maybe we can clarify this in the definition of the field?

One solution could be to have a modifier "onlyProvider" for those functions that are only meant to be called by an actual Provider.

Reply:
Well, `lastPayoutDate` tracks what it says it tracks ;-) (see `payoutFunds` and `_calculateSpentFunds` for details). Only thing is that it is initialized to Agreement creation date for new Agreements, because for new agreements there was still no payout and it is needed to track spent funds. 

Will add some comment about this.

Reply:
Good idea about modifier. Did not know this concept. I will see if it can be used here.